### PR TITLE
Make pointer encoding sound wrt integer address translation (and lift object limits)

### PR DESCRIPTION
**Pull request by tautschnig**
Instead of encoding pointers with the same bit width a pointer on a given
platform has, just widen the bit-blasted encoding to include both the previous
object/offset encoding as well as an (integer) address. The encoding is thus
also trivially be extended to handle larger numbers of objects and offsets of
the same width as the address.

Furthermore clean up the code to encapsulate encoding properly, and make in-code
layout of pointer encoding more natural (it's now object, offset,
integer-address).

Fixes: #436
Fixes: #311
Fixes: #94

### CODE REVIEW

**Comment by tautschnig** [comment]
As far as I'm aware this is now complete (and working) for the SAT back end. The work left to be done is getting #982 merged, adjusting the SMT back ends to do the same. We would then no longer have any limit on the number of objects or the particular addresses that pointers take.

**Comment by martin-cs** [comment]
Another "how alive / urgent is this one" type question?

**Comment by tautschnig** [comment]
It has been among the top elements on my stack for the last couple of weeks; realistically it will stay there until Christmas...

**Comment by martin-cs** [comment]
OK; let me know when having a look at it would be helpful.

**Comment by tautschnig** [comment]
This should be ready for review now. I am aware that possible feedback includes "please refactor into multiple, smaller PRs." All feedback appreciated (including that one). A known point of concern, voiced by @kroening earlier: the pointer flattening is not amenable to incremental use. Ideas to fix that are most welcome!

File: src/analyses/goto_check.cpp [hunk 1]
```diff
@@ -907,23 +904,57 @@ void goto_checkt::pointer_overflow_check(
   if(!enable_pointer_overflow_check)
     return;
 
-  if(expr.id()==ID_plus ||
-     expr.id()==ID_minus)
+  if(expr.id()!=ID_plus &&
+     expr.id()!=ID_minus)
+    return;
```
Comment (thk123) [inline src/analyses/goto_check.cpp]: Nit: since condition is multiple lines can you wrap the return in braces

File: src/analyses/goto_check.cpp [hunk 2]
```diff
@@ -907,23 +904,57 @@ void goto_checkt::pointer_overflow_check(
...
+
+  if(expr.operands().size()>2)
   {
-    if(expr.operands().size()==2)
+    // The overflow checks are binary!
+    // We break these up.
+
+    for(unsigned i=1; i<expr.operands().size(); i++)
```
Comment (thk123) [inline src/analyses/goto_check.cpp]: I needed a pen and paper to fully get what was going on here. I wonder if a comment with an example explaining the calls to pointer_overflow_check would help? 

```C++
// an operand with 4 children, a, b, c, d we:
// first check (a, b)
// then c with a temporary expression with two children: (a, b)
// then d with a temporary expression with three children: (a, b, c)
```

File: src/analyses/goto_check.cpp [hunk 3]
```diff
@@ -907,23 +904,57 @@ void goto_checkt::pointer_overflow_check(
...
-        "overflow",
-        expr.find_source_location(),
-        expr,
-        guard);
+      if(i==1)
+        tmp=expr.op0();
+      else
+      {
+        tmp=expr;
+        tmp.operands().resize(i);
```
Comment (thk123) [inline src/analyses/goto_check.cpp]: You may want to make it explicit here that you are chopping off the last nodes for now to do an intermediate check

**Comment by thk123** [comment]
I've created a TG PR to verify this doesn't break anything, but TBH looks like it won't be a problem: diffblue/test-gen#1322

**Comment by thk123** [comment]
TG PR fails, but for what looks like rebase reasons, can I ask for this PR to be rebased on to current head of develop?

**Comment by smowton** [comment]
Guessing this needs @kroening's opinion?

**Comment by tautschnig** [comment]
This mainly needs work on my end to enable incremental use.

**Comment by tautschnig** [comment]
I have now created #2059, #2060, #2061, #2062, #2063, #2064, #2065, #2066 (and might create at least one more) to place all the preparatory work in separate pull requests that can reasonably be reviewed.

**Comment by peterschrammel** [comment]
Test case from SV-COMP: https://github.com/sosy-lab/sv-benchmarks/blob/master/c/array-memsafety/openbsd_cmemrchr-alloca_true-valid-memsafety_true-termination.i

**Comment by TGWDB** [comment]
@tautschnig This looks interesting but very out of date now. I notice it was close to the top of your stack a few years ago, does that still hold, or has this fallen too far down to keep open? Generally we'd like to close old PRs that are not going to happen.

**Comment by TGWDB** [comment]
Closing due to age and not being actively worked on. Please reopen if you believe this is erroneous and new work towards merging this PR has progressed.

**Pull request closed without merging**